bitcoin/src/taproot/serialized_signature.rs

```diff
@@ -0,0 +1,304 @@
+//! Implements [`SerializedSignature`] and related types.
+//!
+//! Serialized Taproot signatures have the issue that they can have different lengths.
+//! We want to avoid using `Vec` since that would require allocations making the code slower and
+//! unable to run on platforms without allocator. We implement a special type to encapsulate
+//! serialized signatures and since it's a bit more complicated it has its own module.
+
+use core::borrow::Borrow;
+use core::convert::TryFrom;
+use core::{fmt, ops};
+
+pub use into_iter::IntoIter;
+
+use super::Signature;
+use super::SigFromSliceError;
+
+pub(crate) const MAX_LEN: usize = 65; // 64 for sig, 1B sighash flag
+
+/// A serialized Taproot Signature
+#[derive(Copy, Clone)]
+pub struct SerializedSignature {
+    data: [u8; MAX_LEN],
+    len: usize,
+}
```

Comment:
Should this be an enum, with one variant for default sighash and one with the flag byte? This would have the benefit of removing the `len` field and the associated bookkeeping logic surrounding it. We also lose the impossible state when `len > MAX_LEN`, which is not enforced by the type system.

Reply:
Yes, I tried but found we'd have to rewrite `IntoIter` to continue supporting `ReverseIterator` which I was too lazy to do. :) Also this makes it a bit easier to check the impl with `secp256k1`. I can make a followup PR though.

Reply:
I think we should keep it. Agreed that the enum would be conceptually cleaner but symmetry with the ECDSA case also has a lot of value. (e.g. we can copy the new `Ord` impl from this one to the ECDSA one without any changes.)

Reply:
We could in principle rewrite both to never directly access fields and only call `len()` which would be the only one accessing the fields. And also rewrite both iterators. I would absolutely love it but there's so much more important stuff to do now...